From 45c11f8e91a68d09d8904e5f78184f69035a8413 Mon Sep 17 00:00:00 2001
From: Jianxiong Pan <jianxiong.pan@amlogic.com>
Date: Thu, 10 Mar 2022 13:11:46 +0800
Subject: [PATCH 27/95] mm: arm32: add user fault debug support. [1/2]

PD#SWPL-73676

Problem:
need user fault support.

Solution:
add user fault debug support.

Verify:
local.

Change-Id: Ia217724fdee19d8143ca5231baefc310a280d924
Signed-off-by: Jianxiong Pan <jianxiong.pan@amlogic.com>
---
 arch/arm/kernel/process.c | 13 +++++++++++++
 arch/arm/kernel/traps.c   | 22 ++++++++++++++++++++++
 arch/arm/mm/fault.c       | 18 ++++++++++++++++++
 3 files changed, 53 insertions(+)

diff --git a/arch/arm/kernel/process.c b/arch/arm/kernel/process.c
index 0e2d3051741e..9a0ebce3ecf0 100644
--- a/arch/arm/kernel/process.c
+++ b/arch/arm/kernel/process.c
@@ -33,6 +33,9 @@
 #include <asm/mach/time.h>
 #include <asm/tls.h>
 #include <asm/vdso.h>
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+#include <linux/amlogic/user_fault.h>
+#endif
 
 #include "signal.h"
 
@@ -127,6 +130,9 @@ void __show_regs(struct pt_regs *regs)
 
 	printk("PC is at %pS\n", (void *)instruction_pointer(regs));
 	printk("LR is at %pS\n", (void *)regs->ARM_lr);
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+	show_user_fault_info(regs, 0, 0);
+#endif
 	printk("pc : [<%08lx>]    lr : [<%08lx>]    psr: %08lx\n",
 	       regs->ARM_pc, regs->ARM_lr, regs->ARM_cpsr);
 	printk("sp : %08lx  ip : %08lx  fp : %08lx\n",
@@ -148,6 +154,10 @@ void __show_regs(struct pt_regs *regs)
 	buf[3] = flags & PSR_V_BIT ? 'V' : 'v';
 	buf[4] = '\0';
 
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+	show_vmalloc_pfn(regs);
+#endif
+
 #ifndef CONFIG_CPU_V7M
 	{
 		const char *segment;
@@ -187,6 +197,9 @@ void __show_regs(struct pt_regs *regs)
 		printk("Control: %08x%s\n", ctrl, buf);
 	}
 #endif
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+	show_extra_reg_data(regs);
+#endif
 }
 
 void show_regs(struct pt_regs * regs)
diff --git a/arch/arm/kernel/traps.c b/arch/arm/kernel/traps.c
index 91e757bb054e..fb57dc568c6f 100644
--- a/arch/arm/kernel/traps.c
+++ b/arch/arm/kernel/traps.c
@@ -38,6 +38,9 @@
 #include <asm/tls.h>
 #include <asm/system_misc.h>
 #include <asm/opcodes.h>
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+#include <linux/amlogic/user_fault.h>
+#endif
 
 
 static const char *handler[]= {
@@ -51,7 +54,20 @@ static const char *handler[]= {
 void *vectors_page;
 
 #ifdef CONFIG_DEBUG_USER
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+#define USER_DEBUG_UNDEFINED_INSTRUCTION BIT(0)
+#define USER_DEBUG_SYSTEM_CALL           BIT(1)
+#define USER_DEBUG_INVALID_DATA_ABORT    BIT(2)
+#define USER_DEBUG_SIGSEGV_FAULT         BIT(3)
+#define USER_DEBUG_SIGBUS_FAULT          BIT(14)
+unsigned int user_debug = USER_DEBUG_UNDEFINED_INSTRUCTION |
+			  USER_DEBUG_SYSTEM_CALL |
+			  USER_DEBUG_INVALID_DATA_ABORT |
+			  USER_DEBUG_SIGSEGV_FAULT |
+			  USER_DEBUG_SIGBUS_FAULT;
+#else
 unsigned int user_debug;
+#endif
 
 static int __init user_debug_setup(char *str)
 {
@@ -473,6 +489,9 @@ asmlinkage void do_undefinstr(struct pt_regs *regs)
 	if (user_debug & UDBG_UNDEFINED) {
 		pr_info("%s (%d): undefined instruction: pc=%p\n",
 			current->comm, task_pid_nr(current), pc);
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+		show_debug_ratelimited(regs, 0);
+#endif
 		__show_regs(regs);
 		dump_instr(KERN_INFO, regs);
 	}
@@ -517,6 +536,9 @@ asmlinkage void bad_mode(struct pt_regs *regs, int reason)
 {
 	console_verbose();
 
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+	show_debug_ratelimited(regs, 0);
+#endif
 	pr_crit("Bad mode in %s handler detected\n", handler[reason]);
 
 	die("Oops - bad mode", regs, 0);
diff --git a/arch/arm/mm/fault.c b/arch/arm/mm/fault.c
index af5177801fb1..8c92219f6a89 100644
--- a/arch/arm/mm/fault.c
+++ b/arch/arm/mm/fault.c
@@ -21,6 +21,9 @@
 #include <asm/system_misc.h>
 #include <asm/system_info.h>
 #include <asm/tlbflush.h>
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+#include <linux/amlogic/user_fault.h>
+#endif
 
 #include "fault.h"
 
@@ -142,13 +145,28 @@ __do_user_fault(unsigned long addr, unsigned int fsr, unsigned int sig,
 		harden_branch_predictor();
 
 #ifdef CONFIG_DEBUG_USER
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+	if (unhandled_signal(tsk, sig) &&
+	    (((user_debug & UDBG_SEGV) && sig == SIGSEGV) ||
+	     ((user_debug & UDBG_BUS)  && sig == SIGBUS))) {
+#else
 	if (((user_debug & UDBG_SEGV) && (sig == SIGSEGV)) ||
 	    ((user_debug & UDBG_BUS)  && (sig == SIGBUS))) {
+#endif
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+		pr_info("%s: unhandled page fault (%d) at 0x%08lx, code 0x%03x\n",
+			tsk->comm, sig, addr, fsr);
+#else
 		pr_err("8<--- cut here ---\n");
 		pr_err("%s: unhandled page fault (%d) at 0x%08lx, code 0x%03x\n",
 		       tsk->comm, sig, addr, fsr);
+#endif
 		show_pte(KERN_ERR, tsk->mm, addr);
+#ifdef CONFIG_AMLOGIC_USER_FAULT
+		show_debug_ratelimited(regs, 1);
+#else
 		show_regs(regs);
+#endif
 	}
 #endif
 #ifndef CONFIG_KUSER_HELPERS
-- 
2.25.1

